### PR TITLE
feature: New tab page

### PR DESCRIPTION
**Pull request by dnarcese**

### CODE REVIEW

File: newtab-page/newtab.html [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  <link rel="stylesheet" type="text/css" media="all" href="chrome://browser/skin/newtab/newTab.css" />
+</head>
+
+<body xmlns="http://www.w3.org/1999/xhtml" dir="ltr">
+  <div id="newtab-customize-overlay"></div>
+
+  <div id="newtab-vertical-margin">
+    <div id="newtab-margin-top"></div>
+    <div>
+      <iframe src="https://duckduckgo.com/search.html?prefill=Search DuckDuckGo&focus=yes" style="display: block;overflow:hidden;margin: auto;padding:0;width:551px;height:40px;" frameborder="0"></iframe>
```
Comment (bwinton) [inline newtab-page/newtab.html]: I feel like the `&` should probably be an `&amp;` for historical reasons…

File: bundle.js [hunk 2]
```diff
@@ -0,0 +1,2164 @@
+(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
```
Comment (bwinton) [inline bundle.js]: Where did this come from?  It seems kinda big…

File: scripts/newtab.js [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+function onError(error) {
+  console.log(error);
+}
+function loadTopSites(){
+  //display top sites
+  browser.topSites.get().then((topSitesArray) => {
+    var grid = document.querySelector("#newtab-grid");
+    grid.innerHTML = "";
+    for (let topSite of topSitesArray) {
```
Comment (bwinton) [inline scripts/newtab.js:19]: So, on first run, I only see seven or eight top sites, and they're centered which looks kind of odd…  Could we maybe use [a grid layout](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Grid_Layout/Basic_Concepts_of_Grid_Layout) to make things line up better?  (I think `grid-auto-columns` and `grid-auto-rows` will help with this…)

File: scripts/newtab.js [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+      grid.innerHTML +=
+      "<div class='newtab-cell'><div class='newtab-site' draggable='true' type='history'>"+
+          "<a class='newtab-link' title="+topSite.title+" href="+topSite.url+">"+
+            "<span class='newtab-thumbnail thumbnail' style='background-size: cover; background-image: ;'></span>"+
```
Comment (bwinton) [inline scripts/newtab.js]: We should really also add [a contextmenu item](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/contextMenu) to let people copy the thumbnail url.  (It was one of our most requested features!)

File: scripts/newtab.js [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+            "<span class='newtab-title'>"+topSite.title+"</span>"+
+          "</a>"+
+        "</div>"+
+      "</div>"
+    }
+  }, onError);
+}
+function initialize(){
+  fetch(url).then(function(response){
```
Comment (bwinton) [inline scripts/newtab.js]: So, this is the second place we have code like this, if I remember correctly.  Can we extract it into a utility function in another file?

File: scripts/newtab.js [hunk 6]
```diff
@@ -0,0 +1,188 @@
+'use strict'
+
+var placeholders = [];
+var url = 'https://bwinton.github.io/whimsy/thumbnail-gifs.txt';
+var getting = browser.storage.sync.get('newtab');
```
Comment (bwinton) [inline scripts/newtab.js]: Let's save this into a pref, like in the PR I submitted a couple of days ago…

File: scripts/newtab.js [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+        }
+        let rand = Math.floor(Math.random() * placeholders.length);
+        img.style.backgroundImage = "url("+placeholders[rand]+")";
+      })
+    } else if (result.newtab == "hover"){
+      document.querySelectorAll(".newtab-link").forEach(function(link){
+        //set default thumbnail
+        var href = link.href;
+        link.querySelectorAll(".thumbnail").forEach(function(img){
+          //set gifs
```
Comment (bwinton) [inline scripts/newtab.js]: This looks like a lot of code, let's pull it out into a function.

File: scripts/newtab.js [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+      clearInterval(interval);
+    }
+  }
+}
+function onStorageChange(changes, area) {
+  if(changes.newtab.newValue != changes.newtab.oldValue){
+    setThumbnail();
+    //refresh open new tab pages
+    var querying = browser.tabs.query({title: "New Tab"});
+    querying.then((tabs) => {
```
Comment (bwinton) [inline scripts/newtab.js]: Maybe we should only reload the tabs if they're `about:newtab` (or whatever our moz-extension newtab url is)?

**Review by bwinton — CHANGES_REQUESTED** [review]
It works pretty nicely, but there are a few things I'd like to see changed before we merge it.  :smile:

File: bundle.js (hunk 2 above)
Comment (dnarcese) [reply bundle.js]: browserify created that

Comment (bwinton) [reply bundle.js]: I think I'd prefer we included something in a package.json file, and ran browserify as part of a build script.

**Comment by dnarcese** [comment]
Some changes are taking too long, will come back to them at a later time.

Issue #154

File: scripts/newtab.js [hunk 9]
```diff
@@ -0,0 +1,177 @@
+'use strict'
+
+var listener = false;
+var tippyTopSites = require('tippy-top-sites');
+var pref = null;
+
+// initialize('https://bwinton.github.io/whimsy/thumbnail-gifs.txt')
```
Comment (ericawright) [inline scripts/newtab.js]: nit: remove commented code

File: scripts/newtab.js [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+      "<a class='newtab-link' title="+topSite.title+" href="+topSite.url+">"+
+      "<span class='newtab-thumbnail thumbnail' style='background-size: cover; background-position: center; background-image: ;'></span>"+
+      "<span class='newtab-title'>"+topSite.title+"</span>"+
+      "</a>"+
+      "</div>"+
+      "</div>"
+    }
+  }, onError)
+  .then(initialize('https://bwinton.github.io/whimsy/thumbnail-gifs.txt', setThumbnail))
+  //.then(setThumbnail(placeholders));
```
Comment (ericawright) [inline scripts/newtab.js]: also here

**Review by ericawright — APPROVED** [review]

**Pull request merged**